controllers/agents/app/agentapp_controller.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2023 The Primaza Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+
+	"github.com/primaza/primaza/api/v1alpha1"
+	"github.com/primaza/primaza/pkg/primaza/constants"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Agent Application Reconciler reconciles a Agent Application object
+type AgentApplicationReconciler struct {
+	client.Client
+}
+
+func NewAgentApplicationReconciler(mgr ctrl.Manager) *AgentApplicationReconciler {
+	return &AgentApplicationReconciler{
+		Client: mgr.GetClient(),
+	}
+}
+
+const agentappfinalizer = "agentapp.primaza.io/finalizer"
+
+func (r *AgentApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+	l.Info("Reconcile Agent Service Deployment")
+
+	// first, get the agent app deployment
+	agentappdeployment := appsv1.Deployment{}
+	objKey := client.ObjectKey{
+		Name:      constants.ApplicationAgentDeploymentName,
+		Namespace: req.Namespace,
+	}
+	err := r.Get(ctx, objKey, &agentappdeployment)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// it should never happen that this controller does not find itself
+			l.Error(err,
+				"application agent deployment not found, that should be a bug",
+				"expected deployment name", constants.ApplicationAgentDeploymentName)
+			panic(err)
+		}
+		return ctrl.Result{}, err
+	}
+	// TODO: We need to ensure that we do not reconcile service classes once the agent service deployment
+	// is marked for deletion.
+	if agentappdeployment.DeletionTimestamp != nil {
```

Comment:
nit: use the `IsZero()` method on deletion timestamps to check if they're set or not.
```suggestion
	if !agentappdeployment.DeletionTimestamp.IsZero() {
```